### ISSUE TITLE
Error in queue: Applying new propagation method to Neo blocks

### ISSUE THREAD

**Issue by luke-nehemedia** [spicywebau/craft-neo#421]
#### Description
I just did a couple of updates on a clients side. A queue "Applying new propagation method to Neo blocks" was started but persistently fails near the end with the error message: 

> There was a problem getting the parent element.

This is the information given for the queue:
```
{
    "elementType": "benf\\neo\\elements\\Block",
    "criteria": {
        "fieldId": "36"
    },
    "description": "Applying new propagation method to Neo blocks"
}
```

The error-logs (queue.log) show this:
```
2020-12-18 16:20:25 [-][1][-][warning][application] Unable to duplicate “1232” to site 1: Attempting to duplicate an element in an unsupported site.
2020-12-18 16:20:25 [-][1][-][error][craft\errors\UnsupportedSiteException] craft\errors\UnsupportedSiteException: Attempting to duplicate an element in an unsupported site. in /....../craft/vendor/craftcms/cms/src/services/Elements.php:987
Stack trace:
....

2020-12-18 16:21:09 [-][1][-][error][craft\queue\QueueLogBehavior::afterError]  [22103] Applying new propagation method to Neo blocks (attempt: 1) - Error (time: 62.194s): There was a problem getting the parent element.
```

Do you have any idea what the problem is and how to solve it?

#### Other information
- Neo version: 3.6.12
- Craft version: 3.5.17.1
- (if multi-site) What is the affected Neo field's propagation method: The site is multi-site, but I have no idea which field is affected.
- Is eager-loading used? No

**Comment by ttempleton** [comment]
The current Neo version is 2.8.15.1, but if that's the version you have, could you please send your composer.json/lock and database backup to plugins@spicyweb.com.au if possible, and we'll have a look at it.

**Comment by luke-nehemedia** [comment]
Oh yes, I accidentally copied the version of the wrong plugin. 

I just sent you an email with the files.

**Comment by ttempleton** [comment]
Thanks for sending those files, @luke-nehemedia.  The affected Neo block has an `ownerId` which doesn't exist as an element in the database -- it seems that, somehow, the owner entry/category/etc. has been deleted without this Neo block being deleted.  I'll add a migration to delete any Neo blocks with an owner element that doesn't exist, which should be released in the next day.

**Comment by luke-nehemedia** [comment]
Thank you very much!

I have no idea how this happened, but thank you very much for the fix. I am curious to see how it works.

**Comment by ttempleton** [comment]
An update on this: after applying the migration, the error persists; it just involves a different Neo block.  After looking into that further, I believe it's related to issue #363, and found that (at least in this case) the block appears to have been affected by the bug that also caused issue #411, it's associated with two different structures and is using the wrong structure ID when duplicating the block.

I'll need to investigate the best way to fix that, because we're unable to apply the same fix for #363 here, since the `duplicateElement()` call is coming from within the `ApplyNewPropagationMethod` job.  I can't give an exact expected release date for this at the moment, but I'll have it done as soon as possible.  Sorry for the delay with this.

Regarding how the blocks stayed in the database when their owner element no longer existed: I noticed when writing the block hard-deletion migration that block data wasn't actually being deleted from the `neoblocks` table -- which is odd, since it was definitely working for me when I wrote the Neo 2.8.15 migration to soft-delete blocks without an associated structure (at first I was planning to hard-delete them).  My best guess is that something like that has happened when the associated owner element was deleted.  In any case, I've added a fix for the next release, which should make sure those rows in the `neoblocks` table are hard-deleted when they're supposed to be.

**Comment by luke-nehemedia** [comment]
Thank you for all your efforts. 

I have no idea what caused this problem. Maybe the reason was the jump in version numbers. This site has not been updated for a while, so there was quite a jump in the versions. Maybe that caused the problem somehow.

**Comment by ttempleton** [comment]
Just released Neo 2.8.16, where the block propagation fix migration from Neo 2.8.14, that added the queue job causing this error, has been rewritten to apply propagation methods to Neo blocks on a per-block-structure basis rather than a per-field basis.  There are two actions that should be taken before updating (sorry for the inconvenience):

- Cancel the existing failed job
- Delete the row from the `migrations` table where `name` = `m201108_123758_block_propagation_method_fix`

After updating and applying the migrations, you'll have a large number of smaller queue jobs, which is a result of each job focusing on the blocks from a single block structure instead of a field, and should eventually complete with no errors.  I'll close this issue now, but please let us know if any other issues come up with the migration.
